fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/tree/pdagg/PushDownAggregateCollector.java

```diff
@@ -0,0 +1,521 @@
+                    // forbidden push down
+                    return visit(optExpression, context);
+                }
+
+                aggInput.getChild(0).getUsedColumns().getStream().mapToObj(factory::getColumnRef)
+                        .forEach(v -> context.groupBys.put(v, v));
+                aggInput.setChild(0, ConstantOperator.createBoolean(false));
+            }
+        }
+
+        // check has constant aggregate, forbidden
+        if (!context.aggregations.isEmpty() &&
+                context.aggregations.values().stream().allMatch(ScalarOperator::isConstant)) {
+            return visit(optExpression, context);
+        }
+
+        return processChild(optExpression, context);
+    }
+
+    @Override
+    public Void visitLogicalAggregate(OptExpression optExpression, AggregateContext context) {
+        LogicalAggregationOperator aggregate = (LogicalAggregationOperator) optExpression.getOp();
+        // distinct/count* aggregate can't push down
+        if (aggregate.getAggregations().values().stream().anyMatch(c -> c.isDistinct() || c.isCountStar())) {
+            return visit(optExpression, context);
+        }
+
+        // all constant can't push down
+        if (!aggregate.getAggregations().isEmpty() &&
+                aggregate.getAggregations().values().stream().allMatch(ScalarOperator::isConstant)) {
+            return visit(optExpression, context);
+        }
+
+        context = new AggregateContext();
+        context.setAggregator(aggregate);
+        return processChild(optExpression, context);
+    }
+
+    @Override
+    public Void visitLogicalJoin(OptExpression optExpression, AggregateContext context) {
+        if (isInvalid(optExpression, context)) {
+            return visit(optExpression, context);
+        }
+        // constant aggregate can't push down
+        if (!context.aggregations.isEmpty() &&
+                context.aggregations.values().stream().allMatch(ScalarOperator::isConstant)) {
+            return visit(optExpression, context);
+        }
+
+        // split aggregate to left/right child
+        AggregateContext leftContext = splitJoinAggregate(optExpression, context, 0);
+        AggregateContext rightContext = splitJoinAggregate(optExpression, context, 1);
+        process(optExpression.inputAt(0), leftContext);
+        process(optExpression.inputAt(1), rightContext);
+        return null;
+    }
+
+    /*
+     * When aggregation is pushed down to join, it means that the columns on aggregation is from
+     * multi-tables, maybe aggregate columns from left table and group by columns from right table.
+     * We only push down aggregation to child which one support aggregate columns, because aggregate
+     * columns will ignore 1-N join, so there will check:
+     *
+     * 1. all aggregation must come from one child, if none aggregation, can be push down to all children
```

Comment:
A little confused about this sentence. Do you mean: 「all aggregation related columns must come from one child, and it can be push down to both sides if columns is empty, e.g. count(*)?」

Reply:
we can't push down count. e.g. select a, b from A, B group by a, b; -- only group by